File: 1/2021.12.02/743.Network Delay Time/python/solution.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        graph = [[] for _ in range(n + 1)]
        for edge in times:
            graph[edge[0]].append((edge[1], edge[2]))
        dist_to = self.dijkstra(k, graph)
        res = 0
        for i in range(1, len(dist_to)):
            if dist_to[i] == float('inf'):
                return -1
            res = max(res, dist_to[i])
        return res

    def dijkstra(self, start, graph):
        dist_to = [float('inf')] * len(graph)
        dist_to[start] = 0
        heap = []
        heappush(heap, (0, start))
        while heap:
            cur_dist, cur_node = heappop(heap)
            if cur_dist > dist_to[cur_node]:
                continue
            for nex_node, weight in graph[cur_node]:
                nex_dist = cur_dist + weight
                if nex_dist < dist_to[nex_node]:
                    dist_to[nex_node] = nex_dist
                    heappush(heap, (nex_dist, nex_node))
        return dist_to

File: 1/2021.12.02/743.Network Delay Time/java/solution.java
class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {
        List<int[]>[] graph = new LinkedList[n + 1];
        for (int i = 1; i <= n; ++i) {
            graph[i] = new LinkedList<>();
        }
        for (int[] edge : times) {
            int from = edge[0];
            int to = edge[1];
            int weight = edge[2];
            graph[from].add(new int[]{to, weight});
        }
        int[] distTo = dijkstra(k, graph);
        int res = 0;
        for (int i = 1; i <= n; ++i) {
            if (distTo[i] == Integer.MAX_VALUE) {
                return -1;
            }
            res = Math.max(res, distTo[i]);
        }
        return res;
    }

    private int[] dijkstra(int start, List<int[]>[] graph) {
        int[] distTo = new int[graph.length];
        Arrays.fill(distTo, Integer.MAX_VALUE);
        distTo[start] = 0;
        PriorityQueue<State> heap = new PriorityQueue<>((a, b) -> a.distFromStart - b.distFromStart);
        heap.offer(new State(start, 0));
        while (!heap.isEmpty()) {
            State curState = heap.poll();
            int curNodeId = curState.nodeId;
            int curDistFromStart = curState.distFromStart;
            if (curDistFromStart > distTo[curNodeId]) {
                continue;
            }
            for (int[] neighbor : graph[curNodeId]) {
                int nextNodeId = neighbor[0];
                int distToNextNode = curDistFromStart + neighbor[1];
                if (distToNextNode < distTo[nextNodeId]) {
                    distTo[nextNodeId] = distToNextNode;
                    heap.offer(new State(nextNodeId, distToNextNode));
                }
            }
        }
        return distTo;
    }
}

class State {
    int nodeId;
    int distFromStart;

    public State(int nodeId, int distFromStart) {
        this.nodeId = nodeId;
        this.distFromStart = distFromStart;
    }
}
